lib/tree.rb

```diff
@@ -1,62 +1,141 @@
+require "pry"
+
 class TreeNode
   attr_reader :key, :value
   attr_accessor :left, :right
 
-   def initialize(key, val)
+  def initialize(key, val)
     @key = key
     @value = val
     @left = nil
     @right = nil
-   end
+  end
 end
 
 class Tree
   attr_reader :root
+
   def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(Log n)
+  # Space Complexity: O(1)
   def add(key, value)
-    raise NotImplementedError
+    new_node = TreeNode.new(key, value)
+    if !@root
+      @root = new_node
+      return
+    end
+    current = @root
+    while current != nil
+      if key < current.key
+        if current.left
+          current = current.left
+        else
+          current.left = new_node
+          return
+        end
+      else
+        if current.right
+          current = current.right
+        else
+          current.right = new_node
+          return
+        end
+      end
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(Log n)
+  # Space Complexity: O(1)
   def find(key)
-    raise NotImplementedError
+    current = @root
+    while current != nil
+      if key == current.key
+        return current.value
+      elsif key < current.key
+        current = current.left
+      else
+        current = current.right
+      end
+    end
+    return nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(Log n)
+  # Space Complexity: O(n)
   def inorder
```

Comment:
👍 , interesting choice to make an inner function.  That's a very JavaScripty thing to do here.